### ISSUE TITLE
Factor whether or not private vulnerability reporting is enabled into the scorecard

### ISSUE THREAD

**Issue by JasonKeirstead** [ossf/scorecard#2465]
Github has finally added the ability for repository owners to turn on private vulnerability reporting, to make disclosing vulnerabilities in a secure manner easier for all parties involved.

https://docs.github.com/en/code-security/security-advisories/guidance-on-reporting-and-writing/privately-reporting-a-security-vulnerability

The option is not enabled by default right now. 

I believe this should be considered a best practice and factored into the scorecard.

**Comment by david-a-wheeler** [comment]
It's currently in beta, let's make sure it works first. Also: do we want to *mandate* that GitHub projects use it (the implication here)? Some may *want* to report in another way, and the real goal is reporting not the mechanism.

**Comment by katzj** [comment]
Yeah, I think this would be best as counting as success for various parts of the [Security-Policy check](https://github.com/ossf/scorecard/blob/main/docs/checks.md#security-policy)

**Comment by spencerschrock** [comment]
In terms of detecting this, I'm not sure what endpoints we'd have available.

There's one to publish an advisory, but I haven't stumbled across one to query if a project has it enabled.
https://docs.github.com/en/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#privately-report-a-security-vulnerability

**Comment by laurentsimon** [comment]
there's now an API for it https://github.blog/changelog/2024-03-08-check-if-private-vulnerability-reporting-is-enabled-via-rest-api/

Who's interested in implementing this new probe? Could we add this probe to the security policy check maybe?

Thanks @josepalafox for the info!

**Comment by josepalafox** [comment]
And @katecatlin

**Comment by KateCatlin** [comment]
Hi all! As just got shared above, we did indeed [add this endpoint just today](https://github.blog/changelog/2024-03-08-check-if-private-vulnerability-reporting-is-enabled-via-rest-api/) to help you and others check if private vulnerability reporting is enabled at scale. 

Keep us posted on how the implementation goes! This is a really exciting implementation and we completely agree that it's a strong indicator of a repository taking security seriously.

**Comment by spencerschrock** [comment]
> Hi all! As just got shared above, we did indeed [add this endpoint just today](https://github.blog/changelog/2024-03-08-check-if-private-vulnerability-reporting-is-enabled-via-rest-api/) to help you and others check if private vulnerability reporting is enabled at scale.

Just to confirm this is just REST for now, or graphQL as well? I see a `hasVulnerabilityAlertsEnabled ` field in https://docs.github.com/en/graphql/reference/objects#repository, but I think that corresponds to this REST API?https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#check-if-vulnerability-alerts-are-enabled-for-a-repository

**Comment by pnacht** [comment]
> Just to confirm this is just REST for now, or graphQL as well? I see a `hasVulnerabilityAlertsEnabled ` field in https://docs.github.com/en/graphql/reference/objects#repository, but I think that corresponds to this REST API?https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#check-if-vulnerability-alerts-are-enabled-for-a-repository

Yeah, that graphQL field is to check for Dependabot Vulnerability Alerts, not Private Reporting.

**Comment by KateCatlin** [comment]
Yes, confirming this is only on the REST API for now :)

**Comment by Chealer** [comment]
David A. Wheeler wrote:
> [...]
> Also: do we want to _mandate_ that GitHub projects use it (the implication here)?

Of course not. Scorecard is about _scoring_, not _mandating_, and GitHub's new feature is just one of countless tracking mechanisms.

> Some may _want_ to report in another way, and the real goal is reporting not the mechanism.

The goal is _tracking_, not reporting. What mostly matters is the ability to inform enough people with a high (qualification to solve/maliciousness risk) ratio about discovered/suspected vulnerabilities, avoiding to disclose to people with a low ratio. What we need is a metric estimating issue tracking performance. Publishing a maintainer's street address is good, but having a proper ITS is way more reliable and efficient. GitHub's new feature could be even more interesting than any of these in some situations. Even better mechanisms allow reporters to claim their report while controlling its visibility.

If a project which has a public discussion forum has a 3/10 score, it might reach 7/10 when it's had a good but public-only ITS for 5 years. Adding GitHub's feature to that mix could raise to―say―8.

I'd suggest calling the metric _Issue tracking_. I think it would be complicated to evaluate public and private reporting in separate metrics, but it could be given some thought.
Ideally, this would also consider other aspects:

- documentation: Does the website mention where to report issues? Does it give specific guidance on reporting weaknesses? Does the source code contain a REPORTING file?
- incentives: What bounties (monetary rewards) and credits does the project offer reporters?

[🅭🄍](https://www.philippecloutier.com/Common+infrastructure+licensing#its)

**Comment by pnacht** [comment]
My solution here would be to simply include this information in the Security-Policy check.

If the project has a solid security policy that gets a 10/10 already... I honestly don't really care whether they have this feature enabled.
However, if the project doesn't have a security policy but has this feature enabled, then that's already pretty great! I'm spitballing here, but I assume most serious security reports come from security researchers, the vast majority of which are probably aware of this feature and would check for it manually if necessary.

So my suggestion would be:
- Has 10/10 security policy? **10/10**.
- Has no security policy, but has this enabled? **8/10** (slight penalty for slightly lower discoverability)
- Has this enabled and a security policy with a link (likely pointing to this)? **10/10**.

**Comment by diogoteles08** [comment]
I think I agree with Pedro in most points, but I also concern about preventing the following scenario:

A project has a valid Security Policy (currently scoring 10/10), and in its content points to report security vulnerability using GitHub's private vulnerability reporting (not mentioning any other contact method) **but** the feature is not enabled.

I know it might seem rare, but last year GOSST Upstream Team worked to suggest Security Policy for some repositories and we faced that scenario in a relevant amount of times.

Maybe we can use some regex to evaluate if the Security Policy suggest GitHub's tool and, if that's the case, we require that it has it enable in order to get 10/10 ?

**Comment by pnacht** [comment]
That sounds like a good idea, yeah. Dunno if it needs to be done all in one go, though. A PR that simply rewards users for having the feature enabled seems simple enough; a second PR can later add the nuance of "they point to this feature, but don't have it enabled".

**Comment by david-a-wheeler** [comment]
@diogoteles08 - it sounds like "Confidential Issues" are more or less the corresponding feature. Can outsiders determine if they're enabled?

**Comment by diogoteles08** [comment]
I'm not a long-term GitLab user or anything like that, but by my quick research the "Confidential Issues" are enabled by default -- it's a simple checkbox that you can check when your raising any issue. 

More specifically, I think it's true that "if I can create an issue for the project, I can create a confidential issue", so if you can't create an issue for the project I think you can't create a confidential issue. But I don't think this should be a concern, because if a project is not open for new issues, it's likely that it's being really developed and accepting contributions in some other platform (like github or even mailing lists), and has only a mirror on GitLab, for example.

**Comment by david-a-wheeler** [comment]
@diogoteles08 - I think Scorecard should (eventually) check if there's a way to report vulnerabilities. If it's on GitLab, eventually we should check if confidential issues or an email address for reporting vulnerabilities is provided. It's okay if that's not in the initial implementation, it's just good to have that as a long-term objective.

Having no obvious way to report vulnerabilities is a sadly common problem :-(.

**Comment by Bimpebabs** [comment]
I want to take a shot at this. Before I write the code, here's my thinking.

The plan is to add a new field to SecurityPolicyData for private vulnerability reporting. Pointer type so we can tell the difference between enabled, disabled, and unknown. GitHub exposes this through GraphQL as isPrivateVulnerabilityReportingEnabled, so it's just one more field in the existing query. For GitLab and other platforms I'll leave it as unknown for now and handle them in a follow up PR.

Then a new probe called securityPolicyEnablesPrivateReporting, same pattern as the other Security Policy probes. The evaluation change is straightforward. If a repo has a security policy file and PVR is on, that's a 10. Policy file present but PVR unknown gets a 9 so we don't penalize platforms that don't expose the signal. Policy file present but PVR explicitly off gets an 8. No policy file stays the same as today.

I'll add tests for the probe, the evaluation, and the raw data collection with a mocked client. Should have the PR up in the next two weeks.

**Comment by Chealer** [comment]
Thank you @Bimpebabs 

I apologize if I misunderstand your proposal, but if you are saying that a project hosted on GitHub would be penalized compared to a project hosted elsewhere for not using GitHub’s private reporting feature, that seems unfair to me. There are many ways to report issues. A project which requests security reports via a mailing list, without providing alternatives, should get the same score, whether it uses GitHub or not.
